### DISCUSSION TITLE
Reading X11 properties.

### DISCUSSION THREAD

**Discussion by mcnuggetsx20** [qtile/qtile#4278] [Q&A]
How can one go about reading X11 properties set by ```Qtile``` e.g ```_NET_DESKTOP_NAMES``` or ```_NET_NUMBER_OF_DESKTOPS``` with python. It woul be extremely useful to know that.

**Comment by ervinpopescu** [comment]
```py
import os
from libqtile.backend.x11 import xcbq
from xcffib.xproto import GetPropertyReply

def unpack(lst: list):
    for index, el in enumerate(lst):
        if el == b'\x00':
            lst[index] = "\n"
        else:
            lst[index] = el.decode("utf-8")
    return "".join(lst)
display_name = os.environ.get("DISPLAY")
conn = xcbq.Connection(display_name)
_root = conn.default_screen.root
property: GetPropertyReply = _root.get_property("_NET_DESKTOP_NAMES").value
print(unpack(property))
```

This is basically what `qtile` does in `libqtile.backend.x11.window`, but I can't seem to be able to access the `get_property` method defined there.

**Reply by elParaguayo** [reply]
If you have a qtile `Window` object then you need to do `.window.get_property`.

**Reply by ervinpopescu** [reply]
I tried using 
```py
from libqtile.command.client import InteractiveCommandClient
c = InteractiveCommandClient()
c.window.get_property("_NET_DESKTOP_NAMES")
```
But I get:
```
Traceback (most recent call last):
  File "/home/ervin/./debug.py", line 21, in <module>
    print(c.window.get_property("_NET_DESKTOP_NAMES"))
          ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/libqtile/command/client.py", line 236, in __getattr__
    raise SelectError(
libqtile.command.base.SelectError: Not valid child or command, name: get_property, selectors: [('window', None)]
```
It's not decorated with `expose_command` so I suppose that's why.
I also tried to eval it and it didn't work with the same error.

This works:
```py
print(c.window.eval('self.window.get_property("_NET_DESKTOP_NAMES")')[1])
```
but I get `None`. I recall there's something *interesting* going on with `window.window` stuff but I don't know how that works :)

**Reply by elParaguayo** [reply]
The first class is the Qtile window class. The `.window` attribute is an `XWindow` class which has the methods for reading x properties etc.

Are you sure `_NET_DESKTOP_NAMES` is set on the window that you're checking? Does `xprop` show that property?

**Reply by ervinpopescu** [reply]
> Are you sure `_NET_DESKTOP_NAMES` is set on the window that you're checking? Does `xprop` show that property?

I somehow thought xprop didn't give all the properties, since it wasn't among the ones returned by it. But I recognize my mistake :)

> The first class is the Qtile window class. The `.window` attribute is an `XWindow` class which has the methods for reading x properties etc.

Thanks for the explanation. What would be the qtile way of getting the root window? I still wouldn't think there is a need for reading certain properties since most of them are available using qtile's "official" and not so "official" API @mcnuggetsx20. Another way would be using [`wmctrl`](https://pypi.org/project/wmctrl/) and [`ewmh`](https://pypi.org/project/ewmh/0.1.6/)(NB I didn't look into them and I am not endorsing them as safe or whatever else).

[This stackoverflow question](https://stackoverflow.com/questions/9465651/how-can-i-read-an-x-property-using-python-xlib) may be something useful too.

**Reply by elParaguayo** [reply]
Root window is `qtile.core.conn.default_screen.root`. Again, it's an `XWindow` object so you can call `get_property` on it.
